examples/structured_data/feature_preprocessing_layer.py

```diff
@@ -0,0 +1,224 @@
+"""
+Title: Feature Preprocessing with a custom Feature Preprocessor layer.
+Author: [Fernando Nieuwveldt](https://www.linkedin.com/in/fernandonieuwveldt/)
+Date created: 2022/03/21
+Last modified: 2022/03/21
+Description: Feature Preprocessing with a custom Feature Preprocessor layer.
+"""
+"""
+# Feature Preprocessing with a custom Feature Preprocessor layer for an end to end training pipeline.
+"""
+
+"""
+Often for structured data problems we use multiple libraries for preprocessing
+or feature engineering. We might have a full ML training pipeline consisting of
+different libraries for example Pandas for reading data and also feature engineeering,
+sklearn for encoding features for example OneHot encoding and Normalization. The
+estimator might be an sklearn classifier, xgboost or it can for example be a Keras model.
+In the latter case, we would end up with artifacts for feature engineering and encoding
+and different artifacts for the saved model. The pipeline is also disconnected and
+an extra step is needed to feed encoded data to the Keras model. For this step the data
+can be mapped from a dataframe to something like tf.data.Datasets type or numpy array
+before feeding it to a Keras model.
+
+In this example we will implement feature preprocessing natively with
+Keras by implementing a custom Keras layer for all feature preprocessing steps.This layer
+utilizes Keras preprocessing layers. For constructing our model we will use the Functional API.
+"""
+
+"""
+## Setup
+"""
+
+import tensorflow as tf
+from tensorflow.keras.layers import (
+    StringLookup,
+    IntegerLookup,
+    Normalization,
+    Embedding,
+    Discretization,
+    Reshape,
+)
+import pandas as pd
+
+"""
+## Load data
+
+For this example we wil use the income dataset. This dataset contains a mixture of
+numerical and categorical features some with high cardinality.
+"""
+
+CSV_HEADER = [
+    "age",
+    "workclass",
+    "fnlwgt",
+    "education",
+    "education_num",
+    "marital_status",
+    "occupation",
+    "relationship",
+    "race",
+    "gender",
+    "capital_gain",
+    "capital_loss",
+    "hours_per_week",
+    "native_country",
+    "income_bracket",
+]
+
+data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data"
+
+data_frame = pd.read_csv(data_url, header=None, names=CSV_HEADER)
+labels = data_frame.pop("income_bracket")
+labels_binary = 1.0 * (labels == " >50K")
+
+print(f"Train dataset shape: {data_frame.shape}")
+
+"""
+## Split features into different feature groups
```

Comment:
There needs to be some explanation about why you chose to treat certain features as numerical, categorical, to discretize, etc.